### ISSUE TITLE
Memory leak on reads from view, both on_update and to_json

### ISSUE THREAD

**Issue by TimeWanderer** [finos/perspective#1037]
### Steps to Reproduce:
 1. Initialize perspective table with ~10k rows of ~8 columns, half number type, half string type. 
 2. Initialize view that is grouped by half of the columns and aggregated on the other half. 
 3. Start updating table with 2k rows per sec(not inserts, updates with same indices) and monitor memory for a while.
 4. Attach on_update callback with 'cell' mode and monitor memory.

### Expected Result:
Memory footprint in chrome task manager or OS resource monitor(windows task manager in this case) should not increase.

### Actual Result:
Memory stays nearly flat after step 3, but after step 4 there is an increase without any bound. Since there are only updates on the table, and no inserts, it should not increase.

### Environment:
Browser: Chrome Version 81.0.4044.138 (Official Build) (64-bit)
OS: Windows 10

### Additional Context:
JS Heap memory stays more or less the same. Memory footprint in the chrome task manager is increasing. It can also be seen in chrome memory profiles, in which there is a big size delta of ArrayBuffer objects between snapshots at the start and the end. The problem is not only limited to on_update with cell mode. The same thing occurs when there is no attached callbacks, and data is pulled periodically with to_json calls, albeit with gentler increase.

**Comment by sc1f** [comment]
Thank you for the bug report! I have done some repro-ing and it is genuinely a memory leak in `on_update` that has a little to do with how `on_update` is implemented but also how the browser/JS does garbage collection.

### Repro

I reproduced the issue you were seeing [https://gist.github.com/sc1f/42a7933f35d7327912d729701ee65785](here). This gist can be copied into an HTML file and run, after replacing the `data` variable with your own data. The actual dataset doesn't seem to affect how quickly the memory grows or the growth behavior. In this case, I stop the updates after 20 seconds in order to prevent massive growth, as the pattern is clearly observable.

### Results

We can see that the unbounded growth is in the main heap, and not the Web Worker heap (which contains the Emscripten heap):

![Screen Shot 2020-05-22 at 12 46 58 PM](https://user-images.githubusercontent.com/13220267/82715115-b2929b00-9c5f-11ea-9fdd-3adfc9a2163a.png)

With the use of the `updater` function which `console.log()`s the delta that is returned, using the `Performance` tab on Chrome shows memory growth every time the `on_update` callback is fired, and no GC actions (reduction of memory usage) after updates stop (and no GC even after quite some time):

![Screen Shot 2020-05-22 at 1 17 04 PM](https://user-images.githubusercontent.com/13220267/82713916-aeb04a00-9c5a-11ea-9944-d319eac1d979.png)

Manually clicking GC (every 5s in this screenshot) in the chrome debug tools makes no difference - DOM nodes are GC'd but the JS heap is not:

![cell, gc every 5s](https://user-images.githubusercontent.com/13220267/82713965-c7b8fb00-9c5a-11ea-8bac-9ca3dd08069c.png)

Using the row delta, we see that deltas are being generated and GC'd after each iteration. Though usage progressively goes up, I didn't see the JS heap go above 7MB (although this might need more detailed testing, as I figured knowing that the GC was running was enough to set row deltas aside for now):

![row](https://user-images.githubusercontent.com/13220267/82714026-0353c500-9c5b-11ea-92b2-fd656835963b.png)

**However,** when using an updater function that is basically a no-op, the delta is allocated and then immediately GC'd, and the JS heap does not increase unbounded.

<img width="920" alt="Screen Shot 2020-05-22 at 6 39 39 PM" src="https://user-images.githubusercontent.com/13220267/82714196-9ee53580-9c5b-11ea-8986-d4f7c9f85239.png">

I also tried calling `get_step_delta` and `to_json` manually instead of using the `cell` mode in `on_update`, and the memory profiles I got were the same.

### What I think is happening

The difference in memory profiles between the two callbacks makes clear that the delta is not being GC'd automatically, most likely because of how we store callbacks - by _reference_. Because the manual use of GC does not actually deallocate old deltas from the JS heap, it is clear that a reference to each individual delta is being kept somewhere.

To the garbage collector, the delta that is generated is still "reachable" because there is a valid function reference that needs to "reach" the delta, even if semantically it should be garbage collected after the function is called. This seems to be the cause of the issue - in the no-op callback, the generated delta is never referenced so the memory can be allocated and deallocated. But once it's been referenced, it can't be deallocated and thus the memory allocation grows with each iteration. 

I will need to take some time to investigate how to fix this, as it's troubling that the behavior does not replicate itself when using the `row` delta mode - the GC is clearly occurring when using row delta but not cell delta. Additionally, this repro also exposed a related bug - after I call `remove_update`, the memory profile does not change and the GC does not run, which means that even after the initial reference to the update callback is removed from the internal array of callbacks, the memory still can't be GC'd.

**Comment by TimeWanderer** [comment]
@sc1f Have you had a chance to look at memory footprint numbers? From my experience, footprint increases steadily even if you just periodically call to_json. I am not sure as to what memory footprint in the chrome task manager represents. 

I snapshot twice, once at the beginning, and once after I see an obvious increase in memory footprint. When I compared these snapshots, delta sizes were significant. It took much longer to see a significant increase in memory footprint, yet still unacceptable for long-running processes.

**Comment by TimeWanderer** [comment]
Missclicked close and comment button.
:(

**Comment by sc1f** [comment]
Looking at memory footprint (in the Memory tab of the debugger, not in task manager), the JS heap does grow unbounded when the cell mode is used - cell deltas are actually not enabled for pivoted views because the cost of calculating cell deltas (which involves unwrapping the pivots down to their component cells) was considered too high at the time, and so development time was put into row deltas.

I did try calling `to_json` as well, and even with a callback that logged the JSON (thereby referencing it) the JSON was also GC'd. Will edit this with a screenshot - need to dig it up.

**Comment by sc1f** [comment]
Going to mark this issue as closed - I am pretty confident that the issue is not with perspective, but with the `console.log` in the update handlers that I was testing. Because chrome inspects the objects in the console, they can never be GC'd until the console closes/is cleared/the page is closed.

You should be able to test the difference by doing `console.log(JSON.stringify(updated))`, and see that the memory usage pattern corresponds with GC being called after each iteration. Let me know if this helps, or if the issue crops up in another way. Thanks!

**Comment by TimeWanderer** [comment]
I am back with a new test related to this issue. Sorry for not getting back earlier. It has been a crazy year. 

You can see the sample project attached, just run `yarn install` && `yarn start`. This is how it works:

I have two long-running setups with the same perspective table. You should run both setups for about five minutes to see the difference.

- First has a default view, so no grouping. I pull 5000 rows periodically from the view. Memory usage of the worker does not increase.
- Second has a grouping(row_pivots) for each dimension. I pull the same range of data(start_row: 0, end_row: 1000) periodically. You can see worker memory increases as per the linear growth model, i.e steady for a while, then multiplies by some factor.

I do not use the data resolved in to_json returned promise There is only a [weak reference](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakRef) stored, which does not prevent garbage collections. [The weak reference](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakRef) is stored to see when data gets garbage collected.

I used the new [performance.measureMemory](https://web.dev/monitor-total-page-memory-usage/) API to measure worker memory programmatically, which is an experimental feature. I compared the measurement with the memory snapshot from chrome and it seems to be consistent. You need to [enable it in your browser](https://web.dev/monitor-total-page-memory-usage/#using-performance.measurememory()) for this test.

I am not entirely sure if there is a definite memory leak. It may be caused by a number of other factors. I had also suspicions about fragmentation or slow garbage collection which may cause malloc to trigger a memory request for the worker even though there is free memory. I pulled the data with the same parameters every time to eliminate some possibilities. The issue seems to be with grouped views.

[perspective_memleak.zip](https://github.com/finos/perspective/files/5815286/perspective_memleak.zip)

**Comment by TimeWanderer** [comment]
A sample output from memory measurements with timestamp and memory allocated to worker.

![image](https://user-images.githubusercontent.com/4564471/104627793-87563380-56a8-11eb-879a-933d443f1d32.png)

**Comment by sc1f** [comment]
Thanks for the detailed repro! I will pull it down and take a look—I am pretty certain this will be in the C++ part of the codebase, so I will run valgrind against our Python library as well to see if I can narrow it down on a more granular level.

**Comment by TimeWanderer** [comment]
Were you able to make any progress on this?

**Comment by sc1f** [comment]
I've taken a look at the repro and in Python as well - there's definitely a small leak that can be traced back to `get_data` on pivoted contexts, but nothing particularly sticks out so I'm still investigating it.

**Comment by TimeWanderer** [comment]
Hello.

This issue seems to be forgotten. Memory leak grows to gigabytes when tables have many columns. Could you find anything on your last investigation? With 0.7.0 out it seems to be a good time to focus on this issue.

**Comment by texodus** [comment]
> Memory leak grows to gigabytes when tables have many columns.

I don't see evidence of this, however I did find a few smaller (scalar-sized, on the order of your screenshot) leaks which are now fixed.  I've also amped the granularity of our leak tests suite to verify allocation for `to_*()` methods are flat to `< 1` byte/row tolerance against superstore.  If you think there are other methods that leak where they shouldn't, open a new issue and please try to keep it simple w.r.t. repro and expected behavior - it is _a_  _lot_ of work to dissect a two-dozen file example with so many perspective features enabled.

> I do not use the data resolved in to_json returned promise There is only a weak reference stored, which does not prevent garbage collections. The weak reference is stored to see when data gets garbage collected.

Check the MDN docs you link to - it says explicitly that GC of `WeakRef`'s is not guaranteed to _ever_ occur and cannot be used to check allocation behavior.  

> I used the new performance.measureMemory API to measure worker memory programmatically, which is an experimental feature.

This won't work for wasm allocations, they don't show up in the JS heap (in v8 at least).  I've added an `memory_usage()` API which reports the worker/node's `MemoryUsage` object + the real wasm heap size, which is allocated in `16mb * 1.2^N` pages.
